Add CSS styling

botiapa opened the pull request:
Added CSS styling. There was a thing I did not understand when reading your code; What made you implement a custom border-radius implementation? Anyway, I tried replicating the original look as best as I could, but it's not a perfect replica. I'd first like to know whether this is something you want to have in your code before finalizing the PR.

I'd love to hear you're feedback.

ErikReider commented:
> What made you implement a custom border-radius implementation?

I was lazy and didn't want to add gresource logic at the time 😅

> I'd first like to know whether this is something you want to have in your code before finalizing the PR

As long as it's not a pain to maintain, sure :)

ErikReider commented on src/osd_window.rs:
```diff
@@ -36,6 +39,30 @@ impl SwayosdWindow {
 			.style_context()
 			.add_class(&gtk::STYLE_CLASS_OSD.to_string());
 
+		// Set up styling
+		let provider = CssProvider::new();
+		provider.load_from_resource(&format!("{}/style/style.css", DBUS_SERVER_PATH));
+
+		let screen = gtk::prelude::WidgetExt::screen(&window).unwrap();
+		StyleContext::add_provider_for_screen(
+			&screen,
+			&provider,
+			GTK_STYLE_PROVIDER_PRIORITY_APPLICATION as u32,
+		);
+
+		if let Some(user_config_path) = user_style_path() {
+			let user_provider = CssProvider::new();
+			user_provider
+				.load_from_path(&user_config_path)
+				.expect("Failed loading user defined style.css");
+			StyleContext::add_provider_for_screen(
+				&screen,
+				&user_provider,
+				GTK_STYLE_PROVIDER_PRIORITY_USER as u32,
+			);
+			println!("Loaded user defined CSS file");
+		}
+
```
You should be able to do this on first start instead of for each new window

ErikReider commented on src/osd_window.rs:
```diff
@@ -1,22 +1,25 @@
 
 const DISABLED_OPACITY: f64 = 0.5;
 const ICON_SIZE: i32 = 32;
 const WINDOW_MARGIN: i32 = 16;
```
We shouldn't hardcode this, this should be in the CSS file

ErikReider commented on data/style/style.css:
```diff
@@ -0,0 +1,9 @@
+.background {
+    background-color: transparent;
+}
+
+#container {
+    padding: 12px 20px;
+    border-radius: 44px;
+    background-color: @theme_bg_color;
+}
```
This can be simplified to:
```css
window {
    padding: 12px 20px;
    border-radius: 44px;
}
```

ErikReider commented on src/utils.rs:
```diff
@@ -376,3 +378,19 @@ pub fn is_dark_mode(fg: &gdk::RGBA, bg: &gdk::RGBA) -> bool {
 }
+
+pub fn user_style_path() -> Option<String> {
+	match env::var("HOME") {
+		Ok(home) => {
+			let path = Path::new(&home).join(".config/swayosd/style.css");
+			if path.exists() {
+				return path.to_str().map(|s| s.to_string());
+			}
+			None
+		}
+		Err(_) => {
+			println!("$HOME not set, not reading user style.css config");
+			None
+		}
+	}
```
Can be simplified to:
```rust
	let path = user_config_dir().join("swayosd/style.css");
	if path.exists() {
		return path.to_str().map(|s| s.to_string());
	}
	None
```

botiapa commented:
I implemented your proposed changes. I believe the first commit is ready for merge.
I also reworked the progressbar so that it is customizable by CSS as well. There are 2 problems with it currently:
- For some reason on the second init of the progressbar, the inner box fails to change it's width. I'm new to GTK, after debugging it for hours, I couldn't find why. I'm sure it's a trivial fix.
- I have to add the original color scheme back in. Currently the colors are hard coded into the css file, ignoring dark theme settings.

ErikReider commented on src/progressbar.rs:
```diff
@@ -11,90 +12,26 @@ const PROGRESSBAR_HEIGHT: i32 = 6;
-			let height: f64 = f64::from(PROGRESSBAR_HEIGHT);
-			let radius: f64 = height * 0.5;
-
-			// Gets the themes bg and fg colors for the OSD window.
-			// Compares the fg and bg to determine if the theme is a
-			// dark or a light theme.
-			let def_fg = gdk::RGBA::new(0.0, 0.0, 0.0, 1.0);
-			let def_bg = gdk::RGBA::new(1.0, 1.0, 1.0, 1.0);
-			let (fg, bg, dark_mode) = match area.toplevel() {
-				Some(win) => {
-					let style = win.style_context();
-					let fg = style
-						.style_property_for_state(gtk::STYLE_PROPERTY_COLOR, StateFlags::NORMAL)
-						.get::<gdk::RGBA>();
-					let bg = style
-						.style_property_for_state(
-							gtk::STYLE_PROPERTY_BACKGROUND_COLOR,
-							StateFlags::NORMAL,
-						)
-						.get::<gdk::RGBA>();
-					let (fg, bg) = match (fg, bg) {
-						(Ok(fg), Ok(bg)) => (fg, bg),
-						_ => (def_fg, def_bg),
-					};
-					(fg, bg, is_dark_mode(&fg, &bg))
-				}
-				None => (def_fg, def_bg, is_dark_mode(&def_fg, &def_bg)),
-			};
-
-			// Background
-			if dark_mode {
-				ctx.set_source_rgb(fg.red() * 0.5, fg.green() * 0.5, fg.blue() * 0.5);
-			} else {
-				ctx.set_source_rgb(bg.red() * 0.75, bg.green() * 0.75, bg.blue() * 0.75);
-			}
-			ProgressBar::draw_rounded_rectangle(width, height, radius, ctx);
-
-			// Progress
-			ctx.set_source_rgb(fg.red(), fg.green(), fg.blue());
-			ProgressBar::draw_rounded_rectangle(width * fraction, height, radius, ctx);
```
Hard coding the color values in the CSS can make it messy with third-party themes. The current code (in main) uses the themes colors, so still using that by default would be great.

Maybe we can keep this and use it to define a CSS variable on start that the default CSS style can use for the progress bar?

Would you be okay with me pushing changes to your branch?

botiapa replied:
As I said in my comment, I haven't yet tackled making it support third-party themes because the core function does not work. The progressbar works on the first initialization, but it does not work from then on. Please look at my original comment.

Yes of course, feel free to push changes to my branch!

botiapa replied:
Do I need to give you permissions on my repo, or can you push directly from here?

ErikReider replied:
> Do I need to give you permissions on my repo, or can you push directly from here?

I should be able to push without you changing anything :)

ErikReider replied:
Decided to use the themes default foreground color for the progressbar for it to be adaptive :)

ErikReider commented on src/osd_window.rs:
```diff
@@ -271,6 +215,7 @@ impl SwayosdWindow {
 	fn build_progress_widget(&self, fraction: f64) -> crate::progressbar::ProgressBar {
 		cascade! {
 			crate::progressbar::ProgressBar::new(fraction);
```
Can be replaced with a `gtk::progressbar`

ErikReider reviewed (approved):
Thanks for the help! Hopefully you've learnt some GTK while doing this :)
